pkg/config/validator.go

```diff
@@ -364,6 +409,78 @@ func (p *validator) validateAspectRules(rules []*pb.AspectRule, path string, val
 	return numAspects, ce
 }
 
+func (p *validator) validateRules(rules []*pb.Rule, path string) (ce *adapter.ConfigErrors) {
+	for _, rule := range rules {
+		if err := p.validateSelector(rule.GetSelector(), p.descriptorFinder); err != nil {
+			ce = ce.Append(path+":Selector "+rule.GetSelector(), err)
+		}
+
+		path = path + "/" + rule.GetSelector()
+		for idx, aa := range rule.GetActions() {
+			hasError := false
+			hndlr := p.hndlrBldrByName[aa.GetHandler()]
+			if hndlr == nil {
+				hasError = true
+				ce = ce.Appendf(fmt.Sprintf("%s[%d]", path, idx), "handler not specified or is invalid")
+			}
+			for _, instanceName := range aa.GetInstances() {
+				cnstr := p.cnstrByName[instanceName]
+				if cnstr == nil {
```

Comment:
This block can be simplified: you already checked to see if `hndlr` is nil before you started looping. Can we check handler, check the instances, then loop over the valid instances w/ the handler in hand? I think that'll be a lot simpler, at the cost of an extra for loop (oh well; this is out of request path, and the number of things being iterated over is small).